File: Session10/C++/LargestRectangleInHistogram.cpp
https://leetcode.com/problems/largest-rectangle-in-histogram/

class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        
        stack<int> s;
        int n = heights.size();
        int currArea = 0, maxArea = 0, i = 0;
        
        if(n==0) {
            return 0;
        }
        
        while(i<n) {
            while(!s.empty() && heights[i] < heights[s.top()]) {
                int top = s.top();
                s.pop();
                currArea = heights[top] * (s.empty() ? i : i - s.top() - 1);
                maxArea = max(currArea, maxArea);
            }
            s.push(i++);
        }
        
        while(!s.empty()) {
            int top = s.top();
            s.pop();
            currArea = heights[top] * (s.empty() ? i : i - s.top() - 1);
            maxArea = max(currArea, maxArea);
        }
        
        return maxArea;
    }
};

File: Session10/Java/LargestRectangleInHistogram.java
https://leetcode.com/problems/largest-rectangle-in-histogram/

class Solution {
    public int largestRectangleArea(int[] heights) {

        Stack<Integer> stack = new Stack<>();
        int n = heights.length;
        int currArea = 0, maxArea = 0, i = 0;
        
        if (n == 0) {
            return 0;
        }
        
        while (i < n) {
            while (!stack.isEmpty() && heights[i] < heights[stack.peek()]) {
                int top = stack.pop();
                currArea = heights[top] * (stack.isEmpty() ? i : i - stack.peek() - 1);
                maxArea = Math.max(currArea, maxArea);
            }
            stack.push(i++);
        }
        
        while (!stack.isEmpty()) {
            int top = stack.pop();
            currArea = heights[top] * (stack.isEmpty() ? i : i - stack.peek() - 1);
            maxArea = Math.max(currArea, maxArea);
        }
        
        return maxArea;
    }
}